### PR TITLE
add permute transform

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR adds PermuteTransform, which is useful to permute dimensions between normalization flows. `real_vector` constraint is added to correct the domain and codomain of several transforms, including PermuteTransform. It is easy to get things wrong e.g. in case
```
d = dist.MultivariateNormal(loc=np.zeros(3), covariance_matrix=np.identity(3))
x = d.sample()
t  = biject_to(d.support)
unconstrained_value = t.inv(x)
log_prob = d.log_prob(x) + t.log_abs_det_jacobian(unconstrained_value, x)
```
`log_prob` will have shape `3` (which is wrong) if `d.support = real`.

I have checked our current algorithm in hmc/svi/autoguide regarding logdet of transforms. Their logic seems correct so no need to change any of them.

### CODE REVIEW

File: test/test_distributions.py [hunk 1]
```diff
@@ -668,3 +668,40 @@ def test_biject_to(constraint, shape):
 
         assert_allclose(actual, expected, atol=1e-6)
         assert_allclose(actual, -inv_expected, atol=1e-6)
+
+
+# NB: skip transforms which are tested in `test_biject_to`
+@pytest.mark.parametrize('transform, event_shape', [
+    (PermuteTransform(np.array([3, 1, 4, 0, 2])), (5,)),
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply test/test_distributions.py]: Oppss... thanks!

**Comment by fehiepsi** [comment]
> we probably need to add support for event_dim to other transforms like AffineTransform and handle it appropriately in log_abs_det_jacobian, just like we are doing here for IdentityTransform.

Yup, that's the point I have kept thinking about it for a while. Currently, tensorflow handles event dims [explicitly](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/AffineScalar#forward_log_det_jacobian) while pyro/pytorch deals with this event dim logic in algorithm instead. For example, they are at [TransformedDistribution](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/distribution.py#L175), [ComposeTransform](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/constraints.py#L243), or [autoguide](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/autoguide/__init__.py#L177), so log_prob of `TransformedDistribution(mvn_dist, [AffineTransform(0, 1)])` would be fine. However, while writing this reply, I think that the compose of transformed distribution and affine will not work because the current allowable domains of affine are real/greater_than/interval. I will make a simple fix for them but would need to discuss more to find a better solution for general cases.

**Comment by fehiepsi** [comment]
@neerajprad I fixed the issue at AffineTransform. The issue at other transforms are complicated, e.g. if domain of ExpTransform is a real_vector, then its codomain might be a positive vector instead of positive. For now, I don't know what is a good solution to resolve it.

**Comment by neerajprad** [comment]
@fehiepsi - Thanks for fixing that. I more wanted to surface the issue, rather than increase the scope of the current PR. 

The one issue with `real_vector` (in PyTorch as well) is that it conflates value constraints and shape constraints, and doesn't offer a general solution. Maybe a better solution is to add `event_dim` to `log_abs_det_jacobian` as in TensorFlow, and have wrappers like `TransformedDistribution` pass the appropriate value for `event_dim` instead (there might be subtleties that I am missing atm). In any case, this is outside the scope of this PR, but we can try to sketch out a general solution for this rather than handling it piecewise for different combinations of distributions and transforms.
